### PR TITLE
Initial mpMetrics-2.2 implementation

### PR DESCRIPTION
**Pull request by Yushan-Lin**
#9589

### CODE REVIEW

**Comment by Yushan-Lin** [comment]
#build

**Comment by Yushan-Lin** [comment]
#build

**Comment by Yushan-Lin** [comment]
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_qcG9UAAOEeqlbYcQx7mKkA

File: dev/cnf/oss_dependencies.maven [hunk 1]
```diff
@@ -312,6 +312,7 @@ org.eclipse.microprofile.metrics:microprofile-metrics-api:1.0
 org.eclipse.microprofile.metrics:microprofile-metrics-api:1.1.1
 org.eclipse.microprofile.metrics:microprofile-metrics-api:2.0.0
 org.eclipse.microprofile.metrics:microprofile-metrics-api:2.1.0
```
Comment (tevans78) [inline dev/cnf/oss_dependencies.maven]: Do you want to remove the 2.1 dependency if we're not using it?

File: dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0.feature [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0
+visibility=private
+singleton=true
+IBM-Provision-Capability: \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpMetrics-2.2))", \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(|(osgi.identity=com.ibm.websphere.appserver.monitor-1.0)))"
+-bundles=com.ibm.ws.microprofile.metrics.2.0.monitor
+IBM-Install-Policy: when-satisfied
+kind=ga
```
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0.feature]: beta first please

File: dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics2.2-cdi1.2.feature [hunk 3]
```diff
@@ -0,0 +1,11 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.mpMetrics2.2-cdi1.2
+visibility=private
+singleton=true
+IBM-Provision-Capability: \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpMetrics-2.2))", \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(|(osgi.identity=com.ibm.websphere.appserver.cdi-1.2)(osgi.identity=com.ibm.websphere.appserver.cdi-2.0)))"
+-bundles=com.ibm.ws.microprofile.metrics.cdi.2.0
+IBM-Install-Policy: when-satisfied
+kind=ga
```
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics2.2-cdi1.2.feature]: beta first please

File: dev/com.ibm.websphere.appserver.features/visibility/public/mpMetrics-2.2/com.ibm.websphere.appserver.mpMetrics-2.2.feature [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ com.ibm.websphere.appserver.javax.annotation-1.2; ibm.tolerates:=1.3, \
+ com.ibm.websphere.appserver.restHandler-1.0, \
+ com.ibm.websphere.appserver.servlet-3.1; ibm.tolerates:=4.0,\
+ com.ibm.websphere.appserver.mpConfig-1.3; ibm.tolerates:=1.4
+-bundles=com.ibm.ws.require.java8, \
+ com.ibm.ws.microprofile.metrics.common, \
+ com.ibm.ws.microprofile.metrics.2.0, \
+ com.ibm.ws.microprofile.metrics.private, \
+ com.ibm.ws.microprofile.metrics.public
+kind=ga
```
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/public/mpMetrics-2.2/com.ibm.websphere.appserver.mpMetrics-2.2.feature]: beta first please

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+/bin/
```
Comment (tevans78) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/.gitignore]: I don't think you need this gitignore file. These paths are covered at the top level.

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/bnd.bnd [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#*******************************************************************************
+# Copyright (c) 2018 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/bnd.bnd]: 2019

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/bnd.bnd [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+javac.target: 1.8
+
+Bundle-SymbolicName: com.ibm.websphere.org.eclipse.microprofile.metrics.2.2; singleton:=true
+
+Import-Package: \
+  javax.enterprise.util; version="[1.1,3)",\
+  *
+
+Export-Package: \
+  org.eclipse.microprofile.metrics;version=2.1, \
```
Comment (tevans78) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.2.2/bnd.bnd]: I think this should be version 2.2. See https://github.com/eclipse/microprofile-metrics/blob/c1febc5c82fcd90c96aacbc0fa88917b1c16b286/api/src/main/java/org/eclipse/microprofile/metrics/package-info.java#L135

File: dev/com.ibm.ws.microprofile.metrics.2.0.monitor/bnd.bnd [hunk 8]
```diff
@@ -43,6 +43,7 @@ Service-Component: \
 	com.ibm.ws.logging;version=latest,\
 	com.ibm.ws.microprofile.metrics.2.0;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.metrics.2.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.metrics.2.2;version=latest,\
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.metrics.2.0.monitor/bnd.bnd]: On the buildpath you should only pull in the lowest version of the api that you want to be compatible with. So either this new line should be removed or the 2.0 line above should be removed.

File: dev/com.ibm.ws.microprofile.metrics.2.0/bnd.bnd [hunk 9]
```diff
@@ -62,6 +62,7 @@ WS-TraceGroup: METRICS
 	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.metrics.2.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.metrics.2.2;version=latest,\
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.metrics.2.0/bnd.bnd]: On the buildpath you should only pull in the lowest version of the api that you want to be compatible with. So either this new line should be removed or the 2.0 line above should be removed.

**Comment by tevans78** [comment]
Please do not check in binary files: dev/com.ibm.ws.microprofile.metrics.2.2_fat_tck/publish/tckRunner/lib/hamcrest-all-1.3.jar

File: dev/com.ibm.ws.microprofile.metrics.2.2_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/metrics/test/ArquillianLoadableExtension.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.metrics.2.2_fat_tck/publish/tckRunner/tck/src/test/java/com/ibm/ws/microprofile/metrics/test/ArquillianLoadableExtension.java]: 2019

File: dev/com.ibm.ws.microprofile.metrics.cdi.2.0/bnd.bnd [hunk 11]
```diff
@@ -55,6 +55,7 @@ Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
 	com.ibm.ws.kernel.service;version=latest,\
 	com.ibm.ws.microprofile.metrics.2.0;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.metrics.2.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.metrics.2.2;version=latest,\
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.metrics.cdi.2.0/bnd.bnd]: On the buildpath you should only pull in the lowest version of the api that you want to be compatible with. So either this new line should be removed or the 2.0 line above should be removed.

File: dev/cnf/oss_dependencies.maven (hunk 1 above)
Comment (Yushan-Lin) [reply dev/cnf/oss_dependencies.maven]: Will remove all mpMetrics-2.1 references in separate PR

File: dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0.feature (hunk 2 above)
Comment (Yushan-Lin) [reply dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0.feature]: 2.2 reverts back to 2.0, as the change in 2.1 was the problem, our team discussed that there will be no issues to go straight to GA.

Comment (tevans78) [reply dev/com.ibm.websphere.appserver.features/visibility/auto/com.ibm.websphere.appserver.mpMetrics-2.2-monitor-1.0.feature]: you have to go to beta until ALL feature approvals are obtained

File: dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.metrics-2.2.feature [hunk 12]
```diff
@@ -0,0 +1,6 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.org.eclipse.microprofile.metrics-2.2
+singleton=true
+-bundles=com.ibm.websphere.org.eclipse.microprofile.metrics.2.2; location:="dev/api/stable/,lib/"; mavenCoordinates="org.eclipse.microprofile.metrics:microprofile-metrics-api:2.2"
+kind=ga
```
Comment (tevans78) [inline dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.metrics-2.2.feature]: beta first

**Comment by Yushan-Lin** [comment]
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_JdJUwADFEeqlbYcQx7mKkA

**Pull request merged**